### ISSUE TITLE
then() doesn't pass down the context

### ISSUE THREAD

**Issue by x-yuri** [Lakitna/cypress-commands#27]
# Describe the bug
One can't use `this.myAlias` inside `then()` callbacks.

# To Reproduce
```js
    specify('aliases work', function () {
        cy.wrap(1).as('myAlias');
        cy.wrap(2).then(function(subj) {
            expect(this.myAlias).to.eq(1); // undefined
        });
    });

```

# Expected behavior
The test succeeds.

# Versions and such:
 - OS: Arch Linux
 - Cypress version: 3.4.0
 - Cypress browser: Electron 61
 - Cypress-commands version: 0.2.0

# Additional context
A way to fix this would be:

```diff
--- node_modules/cypress-commands/lib/then.js.orig	2019-09-07 10:32:52.749110507 +0300
+++ node_modules/cypress-commands/lib/then.js	2019-09-07 10:33:18.125849439 +0300
@@ -23,7 +23,8 @@
  * @yields {any}
  * @since 0.0.0
  */
-Cypress.Commands.overwrite('then', (originalCommand, subject, fn, options = {}) => {
+Cypress.Commands.overwrite('then', function(originalCommand, subject, fn, options = {}) {
+    const ctx = this;
     if (_.isFunction(options)) {
         // Flip the values of `fn` and `options`
         [fn, options] = [options, fn];
@@ -86,7 +87,7 @@
      */
     async function executeFn() {
         // Execute using the original `then` to not reinvent the wheel
-        return await originalCommand(subject, options, fn)
+        return await originalCommand.call(ctx, subject, options, fn)
             .then((value) => {
                 if (options.log) {
                     consoleProps.Yielded = value;
```
But for one reason or another `originalCommand` [doesn't][1] pass [down][2] the context either.

[1]: https://github.com/cypress-io/cypress/blob/v3.4.1/packages/driver/src/cypress/commands.coffee#L152
[2]: https://github.com/cypress-io/cypress/blob/v3.4.1/packages/driver/src/cypress/commands.coffee#L82

**Comment by Lakitna** [comment]
I did not think about this during the implementation. I'll see if I can add it to 0.3.0

**Comment by Lakitna** [comment]
Oh right, after diving in a bit I remembered.

> But for one reason or another `originalCommand` doesn't pass down the context either.

That's the issue, because of that I can't set the `this` of the overwritten `then()`. That's always the command context (name, prevSubject etc).

There is a workaround though and it has to do with the fat arrow function.

```javascript
it('is a test', function() {
    cy.wrap('fooBar')
        .as('someAlias');

    // Use the fat arrow function syntax here. This will not change `this`, which means
    // it's the `this` of the it-block.
    cy.then(() => {
        expect(this.someAlias).to.equal('fooBar');
    });
});
```

I hope this works for you

**Comment by x-yuri** [comment]
Yes, that can be worked around. And as you might have seen I've filed the issue at [Cypress repository][1].

[1]: https://github.com/cypress-io/cypress/issues/5101

**Comment by Lakitna** [comment]
Alright, I'll keep the issue open until there is clarity on the Cypress repo issue.

**Comment by Lakitna** [comment]
Cypress released this in 6.2.0. I've added tests for this to ensure that it works with both `.then((foo) => {...})` and `.then(function(foo) {...})`.
